core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -371,52 +370,35 @@ public static RecordState forId(byte id) {
      */
     public CompletableFuture<Void> maybeInitialize() {
         log.debug("Maybe initialize share partition: {}-{}", groupId, topicIdPartition);
-        CompletableFuture<Void> future = new CompletableFuture<>();
-        AtomicReference<Optional<Throwable>> futureException = new AtomicReference<>(Optional.empty());
         // Check if the share partition is already initialized.
-        InitializationResult initializationResult = checkInitializationCompletion();
-        if (initializationResult.isComplete()) {
-            if (initializationResult.throwable() != null) {
-                future.completeExceptionally(initializationResult.throwable());
-            } else {
-                future.complete(null);
+        try {
+            if (initializedOrThrowException()) {
+                return CompletableFuture.completedFuture(null);
             }
-            return future;
+        } catch (Exception e) {
+            return CompletableFuture.failedFuture(e);
         }
 
+        // If code reaches here then the share partition is not initialized. Initialize the share partition.
         // All the pending requests should wait to get completed before the share partition is initialized.
         // Attain lock to avoid any concurrent requests to be processed.
         lock.writeLock().lock();
-        boolean shouldFutureBeCompleted = false;
         try {
             // Re-check the state to verify if previous requests has already initialized the share partition.
-            initializationResult = checkInitializationCompletion();
-            if (initializationResult.isComplete()) {
-                if (initializationResult.throwable() != null) {
-                    futureException.set(Optional.of(initializationResult.throwable()));
-                }
-                shouldFutureBeCompleted = true;
-                return future;
+            if (initializedOrThrowException()) {
+                return CompletableFuture.completedFuture(null);
```

Comment:
We should not complete the future while holding the write lock here as well. As mentioned by @AndrewJSchofield in this comment https://github.com/apache/kafka/pull/18053#discussion_r1871156905, this should be completed post releasing the lock, else it may cause problems for the future. Same for the future completion in the `catch` block.

Reply:
So will the call be returned to caller prior finally block execution, if yes then only we should change it else it's not required. So can the future return prior finally block execution?

Reply:
it should right, finally will execute despite a return being present in try and catch.

Reply:
> finally will execute despite a return being present in try and catch

True, I am not understanding what's the problem here. It would be synchronous execution in the same thread. Am I missing something?

Reply:
> it should right

No, it shouldn't. https://docs.oracle.com/javase/specs/jls/se17/html/jls-14.html#jls-14.20.2

_If the finally block completes normally, then the try statement completes normally._

Reply:
If a thread holds the write lock and the returned `CompletableFuture` is registered to complete delayed operations, a deadlock could occur again. However, this may not currently be an issue since we only call `completeDelayedShareFetchRequest` when the future is done.
```java
            CompletableFuture<Void> initializationFuture = sharePartition.maybeInitialize();
            final boolean initialized = initializationFuture.isDone();
            initializationFuture.whenComplete((result, throwable) -> {
                if (!initialized)
                    replicaManager.completeDelayedShareFetchRequest(delayedShareFetchKey);
            });
```

However, I prefer not to hold the write lock while completing the future. Perhaps we can add a new helper method to change the state from empty to initializing.
```java
    private boolean emptyToInitialState() {
        lock.writeLock().lock();
        try {
            if (initializedOrThrowException()) return false;
            partitionState = SharePartitionState.INITIALIZING;
            return true;
        } finally {
            lock.writeLock().unlock();
        }
    }
```
Additionally, the second check can be simplified using the following style:
```java
        // If code reaches here then the share partition is not initialized. Initialize the share partition.
        // All the pending requests should wait to get completed before the share partition is initialized.
        // Attain lock to avoid any concurrent requests to be processed.
        try {
            if (!emptyToInitialState()) return CompletableFuture.completedFuture(null);
        } catch (Exception e) {
            return CompletableFuture.failedFuture(e);
        }
```

Reply:
Your suggestion is better hence I incorporated that.

However I have tested the earlier code as well and as java specification says that without completion of `finally` block the method will not return result from `try`. Unlike earlier, in the problematic code,  where we were completing the `future` in `try`, but now it was a `return` statement.

Below is a test and output.
```
    public CompletableFuture<Void> testFuture() {
        lock.writeLock().lock();
        try {
            System.out.println("testFuture");
            return CompletableFuture.completedFuture(null);
        } finally {
            System.out.println("testFuture finally");
            lock.writeLock().unlock();
            System.out.println("testFuture finally unlocked");
        }
    }
```

```
@RepeatedTest(100)
    public void testFuture() {
        SharePartition sharePartition = SharePartitionBuilder.builder().build();
        sharePartition.testFuture().whenComplete((result, exception) -> {
            System.out.println("Completed");
        });
    }
```

```
testFuture
testFuture finally
testFuture finally unlocked
Completed

Reply:
> now it was a return statement.

Yes, you are right. I did not take account of the "completed" CompletableFuture because I was focused on the relationship between CompletableFuture and the write lock :(